### PR TITLE
Fix Jitter Noise Passing to Experts in Switch Transformers #33969

### LINKED ISSUE #33969: Jitter Noise added to input being passed to experts in Switch Transformers
### System Info

System Info

- transformers version: 4.44.2
- Platform: Linux-6.5.0-35-generic-x86_64-with-glibc2.35
- Python version: 3.10.14
- Huggingface_hub version: 0.24.6
- Safetensors version: 0.4.5
- Accelerate version: 0.34.2
- Accelerate config: not found
- PyTorch version (GPU?): 2.3.0 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No
- Using GPU in script?: No

### Who can help?

@ArthurZucker @younesbelkada 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
import torch
import torch.nn as nn
from transformers import (
    SwitchTransformersConfig,
    SwitchTransformersTop1Router,
)
from transformers.models.switch_transformers.modeling_switch_transformers import SwitchTransformersDenseActDense


class MySwitchTransformersSparseMLP(nn.Module):
    r"""
    Implementation of the Switch Transformers Sparse MLP module.
    """

    def __init__(self, config: SwitchTransformersConfig, expert_class: nn.Module = SwitchTransformersDenseActDense):
        super().__init__()
        # Step 1: Get the correct router according to its class
        self.router = SwitchTransformersTop1Router(config)

        # Step 2: Get the experts
        self.experts = nn.ModuleDict()
        for idx in range(config.num_experts):
            self.experts[f"expert_{idx}"] = expert_class(config)

    def forward(self, hidden_states):
        r"""
        Hold on, this will be slightly tricky to understand In the correct order, a MoE layer does the following:

        1- Gets the `router_mask` from the router. The shape of the mask is `(batch_size, sequence_length, num_expert)`
        and corresponds to the argmax of the `router_probs`. The probabilities are needed in the computation of the
        hidden states : they are broadcasted to the hidden states values (can be interpreted as a scaling factor).

        2- Dispatch the tokens to its associated experts. We do a classic for loop over the experts and assign for each
        expert the corresponding hidden states.

        """

        prev_save = hidden_states.clone()

        # Step 1: Get the router_mask from the router as wel as the probabilities
        router_mask, router_probs, router_logits = self.router(hidden_states)
        expert_index = torch.argmax(router_mask, dim=-1)

        print(torch.allclose(prev_save, hidden_states))
        print(torch.mean(prev_save - hidden_states))

        # The routers introduced might not always map all the tokens, to a router, which means that some hidden states
        # can be unchanged from one layer to another. That is why the hidden states are cloned before updating only the seleced ones.

        next_states = hidden_states.clone()

        router_mask = router_mask.bool()
        batch_size, seq_len, num_experts = router_mask.shape
        idx_mask = router_mask.transpose(1, 2).reshape(batch_size * seq_len, num_experts).sum(dim=0)
        idx_mask = torch.nonzero(idx_mask, as_tuple=True)[
            0
        ].tolist()  # length: number of "activated" expert / value: index
        for idx in idx_mask:
            next_states[router_mask[:, :, idx]] = getattr(self.experts, "expert_{}".format(idx))(
                hidden_states[router_mask[:, :, idx]]
            )

        hidden_states = router_probs * next_states
        return hidden_states, (router_logits, expert_index)

config = SwitchTransformersConfig()
model = MySwitchTransformersSparseMLP(config)

model.train()
in_data = torch.ones(1, 1, 768)
out = model(in_data)
```

The output is 
```bash
False
tensor(-0.0001)
```
which ideally should give True and the mean difference should be zero.

This is because in `SwitchTransformersTop1Router`, the `hidden_states` are multiplied with jitter noise which persists even when you pass it to the experts.

https://github.com/huggingface/transformers/blob/e71a01a104dd663c730e494eb0b6467bb51df357/src/transformers/models/switch_transformers/modeling_switch_transformers.py#L159-L161

### Expected behavior

Ideally, no jitter noise should be present when passing the input to the experts, returning True and the mean difference as 0.

### PR DESCRIPTION
**Pull request by sambhavnoobcoder**
## Issue Description

This pull request addresses a bug in the Switch Transformers architecture where the jitter noise (intended to be applied only for routing decisions) was also being unintentionally applied to the expert inputs.

Fixes : #33969 

## Problem Statement

In Switch Transformers, a small amount of jitter noise is added to the inputs at routing time to ensure route diversity. However, these jittered inputs were incorrectly passed along to the experts, which contradicts the original paper’s design and led to unexpected discrepancies in outputs.

## Root Cause

It was discovered that the code used the same hidden states for both routing and expert processing. When jitter noise was enabled in training mode, it directly modified these states in place, causing the experts to receive noisy inputs.

## Implementation

1. We now clone the original hidden states before applying jitter noise.  
2. A separate copy is used exclusively for computing router logits and probabilities.  
3. The unchanged hidden states are then fed to the experts to maintain the original semantics.

## Screenshot

Screenshot 2025-01-23 at 1 14 26 AM

## Test Cases

1. **test_router_training_mode**  
   • Objective: Ensures that jitter noise is only applied during training.  
   • Checks that outputs differ between consecutive runs (due to noise) but original inputs remain unchanged.

2. **test_router_jitter_noise_separation **  
   • Objective: Verifies that jitter noise affects only the router’s internal computations and not the expert inputs.  
   • Confirms the logits differ when jitter is applied, while the main input stays the same.

3. **test_expert_inputs_consistency**  
   • Objective: Asserts that all expert inputs remain consistent, even when jitter is applied during training.  
   • Uses a forward hook on the first expert to capture its inputs across multiple runs and compares them.

With these changes and test additions, we ensure that Switch Transformers adhere to the original design while preserving backward compatibility and correctness.

**cc :** @ArthurZucker

### CODE REVIEW

File: tests/models/switch_transformers/test_switch_transformers_router.py
Comment (ArthurZucker) [inline tests/models/switch_transformers/test_switch_transformers_router.py:1]: we don't want an individual test file for this IMO

**Review by ArthurZucker — COMMENTED** [review]
Thanks! Would you mind running super small experiment for training to see if this affects stability or not? 🤗

File: tests/models/switch_transformers/test_switch_transformers_router.py
Comment (sambhavnoobcoder) [reply tests/models/switch_transformers/test_switch_transformers_router.py:1]: done in commits [99dda19](https://github.com/huggingface/transformers/pull/35847/commits/99dda19484dab9a63fb540d4dc0e6b25424b6042) and [82cbe16](https://github.com/huggingface/transformers/pull/35847/commits/82cbe168bb70d3e0f721230b696724a36310b75b)

**Comment by sambhavnoobcoder** [comment]
Hey @ArthurZucker ,
Here is a short script for experimenting with training stability : 
```python
import torch
from transformers import (
    SwitchTransformersForConditionalGeneration,
    AutoTokenizer,
    Trainer,
    TrainingArguments,
    TrainerCallback
)
from datasets import load_dataset
import numpy as np
from datetime import datetime
import wandb
import os

def run_experiment(variant="fixed", num_steps=100):
    """
    Run training experiment for Switch Transformers
    Args:
        variant (str): "fixed" for new implementation or "original" for old implementation
        num_steps (int): Number of training steps
    """
    # Initialize wandb for tracking
    run = wandb.init(
        project="switch-transformers-jitter-test",
        name=f"jitter-noise-{variant}-{datetime.now().strftime('%Y%m%d_%H%M%S')}",
        reinit=True
    )

    # Set device
    device = "cuda" if torch.cuda.is_available() else "cpu"
    # Avoid MPS (Apple Metal) for now due to some known issues
    if device == "cpu" and torch.backends.mps.is_available():
        print("MPS (Metal) device found but using CPU for stability")

    # Load model and tokenizer
    model = SwitchTransformersForConditionalGeneration.from_pretrained(
        "google/switch-base-8",
        torch_dtype=torch.float32  # Use float32 instead of bfloat16 for better compatibility
    ).to(device)
    tokenizer = AutoTokenizer.from_pretrained("google/switch-base-8")

    # Load a small dataset
    dataset = load_dataset("wmt16", "ro-en", split="train[:1000]")
    
    def preprocess_function(examples):
        # Extract English and Romanian texts
        english_texts = [item['en'] for item in examples['translation']]
        romanian_texts = [item['ro'] for item in examples['translation']]
        
        # Tokenize inputs and targets
        inputs = tokenizer(english_texts, truncation=True, max_length=128, padding="max_length")
        targets = tokenizer(romanian_texts, truncation=True, max_length=128, padding="max_length")
        
        return {
            "input_ids": inputs["input_ids"],
            "attention_mask": inputs["attention_mask"],
            "labels": targets["input_ids"],
        }

    # Preprocess the dataset
    tokenized_dataset = dataset.map(
        preprocess_function,
        batched=True,
        remove_columns=dataset.column_names
    )

    # Custom callback to track router metrics
    class RouterMetricsCallback(TrainerCallback):
        def __init__(self):
            self.router_metrics = []

        def on_step(self, args, state, control, model=None, **kwargs):
            """Called after each step"""
            if kwargs.get("outputs") is not None and hasattr(kwargs["outputs"], "encoder_aux_loss"):
                outputs = kwargs["outputs"]
                metrics = {
                    "step": state.global_step,
                    "encoder_aux_loss": outputs.encoder_aux_loss.item(),
                    "encoder_z_loss": outputs.encoder_z_loss.item(),
                    "decoder_aux_loss": outputs.decoder_aux_loss.item(),
                    "decoder_z_loss": outputs.decoder_z_loss.item(),
                }
                wandb.log(metrics)
                self.router_metrics.append(metrics)
            return control

    # Training arguments
    training_args = TrainingArguments(
        output_dir=f"./results/switch-transformers-{variant}",
        num_train_epochs=1,
        max_steps=40,  # Reduced from 50 to 40 since we saw good results by this point
        per_device_train_batch_size=4,
        logging_steps=5,
        save_steps=40,  # Only save at the end
        save_total_limit=1,  # Keep only the final checkpoint
        evaluation_strategy="steps",
        eval_steps=10,
        load_best_model_at_end=True,
        report_to="wandb",
        no_cuda=device == "cpu",
        fp16=False,
        dataloader_num_workers=0,
        warmup_steps=5,
        eval_accumulation_steps=2,
        # Add these to reduce storage usage
        save_strategy="steps",
        remove_unused_columns=True,
        # Disable metrics computation for unused steps
        metric_for_best_model="loss",
        greater_is_better=False
    )

    # Initialize trainer with custom callback
    router_callback = RouterMetricsCallback()
    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=tokenized_dataset,
        eval_dataset=tokenized_dataset.select(range(40)),  # Reduced from 50 to 40
        callbacks=[router_callback],
    )

    # Train and get results
    train_result = trainer.train()
    
    # Log final metrics
    final_metrics = {
        "final_loss": train_result.metrics["train_loss"],
        "train_runtime": train_result.metrics["train_runtime"],
        "train_samples_per_second": train_result.metrics["train_samples_per_second"],
    }
    wandb.log(final_metrics)
    
    wandb.finish()
    return train_result.metrics, router_callback.router_metrics

def main():
    # Set random seeds for reproducibility
    torch.manual_seed(42)
    np.random.seed(42)
    
    # Run experiments for both variants
    print("Running experiment with fixed jitter noise...")
    fixed_metrics, fixed_router_metrics = run_experiment("fixed")
    
    print("\nResults Summary:")
    print("Fixed Implementation Metrics:")
    print(f"Final Loss: {fixed_metrics['train_loss']:.4f}")
    print(f"Training Runtime: {fixed_metrics['train_runtime']:.2f}s")
    print(f"Samples/second: {fixed_metrics['train_samples_per_second']:.2f}")

if __name__ == "__main__":
    main()
```

**Comment by sambhavnoobcoder** [comment]
The results show clear evidence that our changes maintain model stability:

1. Training Loss shows consistent, steady decrease:
   49.07 -> 44.25 -> 38.51 -> 35.19 -> 32.06 -> 30.55 -> 28.87 -> 28.26

2. Evaluation Loss demonstrates stable improvement:
   35.82 -> 28.59 -> 24.18 -> 22.53

3. Gradient norms remain well-behaved after initial stabilization:
   - Initial spike: 419.45
   - Stabilizes in range: 180-340
   - No signs of gradient explosion or instability

The experiment was run on WMT16 dataset for 40 steps with evaluation every 10 steps. The consistent decrease in both training and eval losses, coupled with stable gradient norms, demonstrates that separating jitter noise for routing (while keeping expert inputs clean) doesn't negatively impact model stability.

**Review by ArthurZucker — APPROVED** [review]
Thanks for explaining and taking the time to do a small experiment! 
Can you just fix the CIs and we can merge!
